File: problems/前缀和/974.和可被K整除的子数组.cpp
#include <iostream>
#include <vector>
using namespace std;

int subarraysDivByK(vector<int>& nums, int k) {
    vector<int> modCount(k, 0);
    modCount[0] = 1;
    int prefixMod = 0, count = 0;
    for (int num : nums) {
        prefixMod = (prefixMod + num) % k;
        if (prefixMod < 0) prefixMod += k;
        count += modCount[prefixMod];
        modCount[prefixMod]++;
    }
    return count;
}

int main() {
    // 测试用例
    vector<int> nums1 = {4, 5, 0, -2, -3, 1};
    cout << subarraysDivByK(nums1, 5) << endl;
    // 预期: 7

    vector<int> nums2 = {5};
    cout << subarraysDivByK(nums2, 9) << endl;
    // 预期: 0

    vector<int> nums3 = {-1, 2, 9};
    cout << subarraysDivByK(nums3, 2) << endl;
    // 预期: 2

    return 0;
}

File: problems/前缀和/974.和可被K整除的子数组.py
#!/usr/bin/env python3
"""
974. 和可被K整除的子数组
"""
def subarrays_div_by_k(nums: list[int], k: int) -> int:
    prefix_mod = 0
    mod_count = [0] * k
    mod_count[0] = 1
    count = 0
    for num in nums:
        prefix_mod = (prefix_mod + num) % k
        if prefix_mod < 0:
            prefix_mod += k
        count += mod_count[prefix_mod]
        mod_count[prefix_mod] += 1
    return count

if __name__ == "__main__":
    # 测试用例
    print(subarrays_div_by_k([4, 5, 0, -2, -3, 1], 5))
    # 预期: 7
    print(subarrays_div_by_k([5], 9))
    # 预期: 0
    print(subarrays_div_by_k([-1, 2, 9], 2))
    # 预期: 2
